### ISSUE TITLE
Computer locks up when locking screen

### ISSUE THREAD

**Issue by thopri** [cinatic/taskwhisperer#88]
When installed, this extension results in a freeze in the computer of up to 5 mins when locking display. Computer completely unresponsive until lock screen appears and is fine until locking again, also had the freeze when disabling the extension from the extensions program.

Using:  Gnome 40 and Fedora 34 Beta

This delay in locking makes it unfeasible to use, which is a shame as I do like the extension and find it nice to use.

**Comment by cinatic** [comment]
Hi and thanks for the report,

Could you provide logs maybe 20 min before and after locking?

`sudo journalctl /usr/bin/gnome-shell`
Or even of the whole system if you like `sudo journalctl -xe`

You can also send it to me directly florijanh@gmail.com

If extension is disabled this shouldn't happen at all as the code is not loaded, however maybe something in the extension causes this.

**Comment by cinatic** [comment]
i just installed fedora 34 in a vm and give it a try, i don't get the freeze however i found some errors when destroying event is triggered. When you say you have the problem even when disabling the extension, then it maybe because the destroy is not working properly. 

Please try the following

<details>
<summary>

```
1. replace ~/.local/share/gnome-shell/extensions/taskwhisperer-extension@infinicode.de/extension.js with the code below.
2. log out / log in
3. try to reproduce
```

</summary>

```
/* jshint esnext:true */
/*
 *
 *  GNOME Shell Extension for the great Taskwarrior application
 *
 * Copyright (C) 2021
 *     Florijan Hamzic <fh @ infinicode.de>
 *
 * This file is part of taskwhisperer-extension.
 *
 * taskwhisperer-extension is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * taskwhisperer-extension is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with taskwhisperer-extension.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

const { Clutter, GObject, St } = imports.gi

const ExtensionUtils = imports.misc.extensionUtils
const Me = ExtensionUtils.getCurrentExtension()

const { MenuItem } = Me.imports.components.panel.menuItem
const { ScreenWrapper } = Me.imports.components.screenWrapper.screenWrapper
const { EventHandler } = Me.imports.helpers.eventHandler
const { initTranslations } = Me.imports.helpers.translations
const { Settings } = Me.imports.helpers.settings

const Gettext = imports.gettext.domain('taskwhisperer@infinicode.de')
const _ = Gettext.gettext

const Main = imports.ui.main
const PanelMenu = imports.ui.panelMenu

const MenuPosition = {
  CENTER: 0,
  RIGHT: 1,
  LEFT: 2
}

let TaskWhispererMenuButton = GObject.registerClass(class TaskWhispererMenuButton extends PanelMenu.Button {
  _init () {
    this._currentPanelPosition = null
    this._settingsChangedId = null

    // Panel menu item - the current class
    let menuAlignment = 0.25

    if (Clutter.get_default_text_direction() == Clutter.TextDirection.RTL) {
      menuAlignment = 1.0 - menuAlignment
    }

    super._init(menuAlignment, _('taskwarrior'))
    this.add_style_class_name('taskwhisperer-extension')

    this.add_child(new MenuItem())

    const bin = new St.Widget({ style_class: 'taskwhisperer-extension' })
    bin._delegate = this
    this.menu.box.add_child(bin)

    this._screenWrapper = new ScreenWrapper()
    bin.add_child(this._screenWrapper)

    // Bind events
    EventHandler.connect('hide-panel', () => this.menu.close())
    this._settingsChangedId = Settings.connect('changed', (changedValue, changedKey) => this._sync(changedValue, changedKey))

    this.menu.connect('destroy', this._destroyExtension.bind(this))
    this.menu.connect('open-state-changed', (menu, isOpen) => {
      EventHandler.emit('open-state-changed', { isOpen })
    })

    this._sync()
  }

  _sync (changedValue, changedKey) {
    this.checkPositionInPanel()
  }

  checkPositionInPanel () {
    const newPosition = Settings.position_in_panel

    if (this._currentPanelPosition === newPosition) {
      return
    }

    this.get_parent().remove_actor(this.actor)

    switch (this._currentPanelPosition) {
      case MenuPosition.LEFT:
        Main.panel._leftBox.remove_actor(this.actor)
        break
      case MenuPosition.CENTER:
        Main.panel._centerBox.remove_actor(this.actor)
        break
      case MenuPosition.RIGHT:
        Main.panel._rightBox.remove_actor(this.actor)
        break
    }

    let children = null
    switch (newPosition) {
      case MenuPosition.LEFT:
        children = Main.panel._leftBox.get_children()
        Main.panel._leftBox.insert_child_at_index(this.actor, children.length)
        break
      case MenuPosition.CENTER:
        children = Main.panel._centerBox.get_children()
        Main.panel._centerBox.insert_child_at_index(this.actor, children.length)
        break
      case MenuPosition.RIGHT:
        children = Main.panel._rightBox.get_children()
        Main.panel._rightBox.insert_child_at_index(this.actor, 0)
        break
    }

    this._currentPanelPosition = newPosition
  }

  _destroyExtension () {
    if (this._settingsChangedId) {
      Settings.disconnect(this._settingsChangedId)
    }
  }
})

var taskWhispererMenu

function init (extensionMeta) {
  initTranslations()
}

function enable () {
  taskWhispererMenu = new TaskWhispererMenuButton()
  Main.panel.addToStatusArea('taskWhispererMenu', taskWhispererMenu)
}

function disable () {
  taskWhispererMenu.destroy()
}

```

</details>

**Comment by thopri** [comment]
Hi Florijan,

Sorry I didn't write that properly the freeze occurs when clicking the deactivate switch, when the computer unfreezes it seems to be ok even with the extension installed. Although I am not sure as I uninstalled the extension at that point. I wasn't sure what was causing the issue and uninstalled yours first since it was the last one to be installed.

I will reinstall and try the fix you suggest, I have got the log you requested, doesnt seem like there is much but will email it to you.

**Comment by thopri** [comment]
Here is the log output,

Will apply the fix you mention.

Many Thanks

Tom


Dr Thomas Prime
Applied Modeller
UK National Oceanography Centre, European Way, Southampton SO14 3ZH

E:  ***@***.******@***.***>
noc.ac.uk<http://noc.ac.uk/> | @NOCnews<https://twitter.com/nocnews> | Subscribe to NOCmail e-alerts<https://public.govdelivery.com/accounts/UKNOC/subscriber/new>


DISCLAIMER:  This communication (including any attachments) is intended for the addressee only and may contain confidential, privileged or copyright material. It may not be relied upon or disclosed to any other person without the consent of the National Oceanography Centre. In the event that this communication has incorrectly been addressed to you, it is likewise subject to protection under confidentiality and should not be disclosed.  In either event if you have received this communication in error, please permanently delete it from your system and promptly notify the sender. You should not copy it or use it for any purpose nor disclose or distribute its contents to any other person.

________________________________

**Comment by thopri** [comment]
Sorry just remembered I should state I have pop shell installed (from POP!_OS) I had to build this from source to get it to work on Gnome 40, while this isn't causing the issue. It may be interfering with your extentsion.

So I will try it with that deactivated as well.

**Comment by thopri** [comment]
Freeze still happens with POP shell deactivated.

**Comment by thopri** [comment]
Wahey! your fix has worked computer no longer sits frozen  for 5 mins with its fans blazing. Nice one

**Comment by cinatic** [comment]
glad it finally worked and thanks for the support. i guess calling destroy() in the onDestroy() event can cause a recursive chain depending maybe on downstream code (idk), strange also that i never fall over this bug. i will roll out this soon in a new version

**Comment by thopri** [comment]
cool, am glad its a nice extension.

**Comment by cinatic** [comment]
fixed with #89
